python/kaitai/compress/core/__init__.py

```diff
@@ -0,0 +1,22 @@
+class KaitaiProcessorContext:
+       def __call__(self, slce:slice):
+           raise NotImplementedException("Please implement process")
+
+class ProcessorContextStub(KaitaiProcessorContext):
+    def __init__(self, obj, meth, data):
+        self.obj=obj
+        self.meth=meth
+        self.data=meth(data)
+    def __call__(seld, slc:slice):
+        return self.data[slc]
+
+class KaitaiProcessor:
+    """The base processor class"""
+    def __init__(self, *args, **kwargs):
+        raise NotImplementedException("Please implement __init__")
+    def process(self, data:(bytes, bytearray), *args, **kwargs) -> KaitaiProcessorContext:
+        raise NotImplementedException("Please implement process")
+    def unprocess(self, data:(bytes, bytearray), *args, **kwargs ) -> KaitaiProcessorContext:
+        raise NotImplementedException(self.__class__.__name__+" processing is not invertable")
+    def getArgs(self, data:(bytes, bytearray) ):
+        raise NotImplementedException("Cannot get args of "+self.__class__.__name__)
```

Comment:
If this is a new API proposal:

1. It should have been discussed separately, and not bundled all together with one big PR.
2. I don't see how it's radically different from existing API. The only major reason to update APIs I can think of is support of streaming transforms, and that will likely involve wrapping around KaitaiStream implementation.
3. This uses Python 3.5+ things, which is unlikely a good idea for a very general-purpose library.
4. `getArgs` probably needs to be spelled as `get_args`, [as per PEP8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names).

Reply:
>don't see how it's radically different from existing API. The only major reason to update APIs I can think of is support of streaming transforms, and that will likely involve wrapping around KaitaiStream implementation.

Yes, this was the main motivation - allow seekable transformations.

>This uses Python 3.5+ things, which is unlikely a good idea for a very general-purpose library.

I don't see where. If you mean type annotations, they are available in 3.4 too, and probablynin earlier versions.

>getArgs probably needs to be spelled as get_args, as per PEP8.

OK